src/Gui/NotificationArea.cpp

```diff
@@ -0,0 +1,993 @@
+    auto current_thread = QThread::currentThread();
+
+    if(timer_thread == current_thread)
+        pImp->inhibitTimer.start(pImp->inhibitNotificationTime);
+}
+
+bool NotificationArea::confirmationRequired(Base::LogStyle level)
+{
+    auto userInitiatedRestore =
+        Application::Instance->testStatus(Gui::Application::UserInitiatedOpenDocument);
+
+    return (level == Base::LogStyle::Critical && userInitiatedRestore
+            && pImp->requireConfirmationCriticalMessageDuringRestoring);
+}
+
+void NotificationArea::showConfirmationDialog(const QString& notifiername, const QString& message)
+{
+    auto confirmMsg = QObject::tr("Notifier: ") + notifiername + QStringLiteral("\n\n") + message
+        + QStringLiteral("\n\n")
+        + QObject::tr("Do you want to skip confirmation of further critical message notifications "
+                      "while loading the file?");
+
+    auto button = QMessageBox::critical(getMainWindow()->activeWindow(),
+                                        QObject::tr("Critical Message"),
+                                        confirmMsg,
+                                        QMessageBox::Yes | QMessageBox::No,
+                                        QMessageBox::No);
+
+    if (button == QMessageBox::Yes)
+        pImp->requireConfirmationCriticalMessageDuringRestoring = false;
+}
+
+void NotificationArea::showInNotificationArea()
+{
+    std::lock_guard<std::mutex> g(
+        pImp->mutexNotification);// guard to avoid modifying the notification list and indices while
+                                 // creating the tooltip
+
+    NotificationsAction* na = static_cast<NotificationsAction*>(pImp->notificationaction);
+
+    if (!NotificationBox::isVisible()) {
+        // The Notification Box may have been closed (by the user by popping it out or by left mouse
+        // button) ensure that old notifications are not shown again, even if the timer has not
+        // lapsed
+        int i = 0;
+        while (i < na->count() && static_cast<NotificationItem*>(na->getItem(i))->notifying) {
+            NotificationItem* item = static_cast<NotificationItem*>(na->getItem(i));
+
+            if (item->shown) {
+                item->notifying = false;
+                item->shown = false;
+            }
+
+            i++;
+        }
+    }
+
+    auto currentlyshown = na->getShownCount();
+
+    // If we cannot show more messages, we do no need to update the non-intrusive notification
+    if (currentlyshown < pImp->maxOpenNotifications) {
+        // There is space for at least one more notification
+        // We update the message with the most recent up to maxOpenNotifications
+
+        QString msgw =
+            QString::fromLatin1(
+                "<style>p { margin: 0 0 0 0 } td { padding: 0 15px }</style>                     \
+        <p style='white-space:nowrap'>                                                                                      \
```

Comment:
@abdullahtahiriyo 
While looking at issue https://github.com/FreeCAD/FreeCAD/issues/8860 I got a very long error message. Because the HTML code uses nowrap the label occupied the whole width of the screen and only a part of the message was shown. Wouldn't it make more sense to use `<p style='white-space:wrap'>`?

Reply:
I remember there were visualisation problems when not using "nowrap".

I will investigate this further.

Reply:
Qt offers for white-space normal, pre, nowrap, pre-wrap.

With normal, and pre-wrap short messages become strongly multiline. Example pre-wrap:
![image](https://user-images.githubusercontent.com/5249162/224752402-a691d9b5-1ed9-4a00-8e62-0dce787738b3.png)

Example normal:
![image](https://user-images.githubusercontent.com/5249162/224753532-ed9293c3-bdad-4849-a082-ab3d71061e5b.png)

With pre, the notification gets too big:
![image](https://user-images.githubusercontent.com/5249162/224752731-a1f8e6ee-6e48-438e-a6ee-6e6f4ab11a3e.png)

nowrap is the one providing the best visual results (except when a message is too long):
![image](https://user-images.githubusercontent.com/5249162/224753287-2ad1ded9-cca2-45c7-8bbd-e97d401954ec.png)
![image](https://user-images.githubusercontent.com/5249162/224753325-79da6eae-2b93-4de4-810b-271d1a4868f7.png)
![image](https://user-images.githubusercontent.com/5249162/224753410-b65d1886-8711-4c45-80dc-7df29336e79e.png)

There are consequences from wrapping. The vertical space occupied by accumulated messages is controlled by the maximum number of simultaneous non-intrusive communications (defaults to 15). However, if message extend over different lines, then the vertical space can be higher.

I am still looking for a solution:
- The best looking nowrap does not have a fixed length for the qlabel, which adapts to the text. That is why fixing the size (width), which would also need to take into account the screen size, has some limitations.
- Then I thought of providing a user preference for the maximum text length, so that the user can configure it. When the maximum text length is surpassed, the notificationbox temporarily changes mode, or somehow introduces </br>... it does not look very nice though...

I keep thinking...

Reply:
In the Python console I entered:
```
Err("""Running the Python command 'MeshPart_CreateFlatMesh' failed:
Traceback (most recent call last):
  File "/Applications/FreeCAD.app/Contents/Resources/Mod/MeshPart/MeshFlatteningCommand.py", line 58, in Activated
    boundaries = flattener.getFlatBoundaryNodes()

Unable to convert function return value to a Python type! The signature was
	(self: flatmesh.FaceUnwrapper) -> std::__1::vector<Eigen::Matrix<double, -1, 3, 0, -1, 3>, std::__1::allocator<Eigen::Matrix<double, -1, 3, 0, -1, 3> > >

Did you forget to `#include <pybind11/stl.h>`? Or <pybind11/complex.h>,
<pybind11/functional.h>, <pybind11/chrono.h>, etc. Some automatic
conversions are optional and require extra headers to be included
when compiling your pybind11 module.""")
```
and got this very wide label. In the mentioned line I simply removed the "no" from nowrap and then for this case the label text looked more useful to me:

But if there is no better alternative than nowrap for the general case then it would be good to not let the label occupy the whole width of the screen but maybe up to 30 % or so.

Reply:
I will check with "wrap". I did not found that option documented:
https://doc.qt.io/qt-6/richtext-html-subset.html
https://developer.mozilla.org/en-US/docs/Web/CSS/white-space

Reply:
I will push a PR with:
- Fixed width notification, which is configurable from preferences.
- Normal white-space, which means that the text will wrap to fit the contents.
- Enforcing via <BR> of any \n existing in the message. This preserves the intended structure of the message (unless the message is too big for the fixed width).

Then your Error message renders like this in my screen:
![image](https://user-images.githubusercontent.com/5249162/225717142-3b2e45b9-b122-4661-b02b-559de8c42d9a.png)

Reply:
But then looking at it, I realise that it needs some escaping... will come back...

Reply:
Better now:
![image](https://user-images.githubusercontent.com/5249162/225723873-bfdb4f3e-019c-4dc1-b5a0-caa781d88c54.png)

Reply:
https://github.com/FreeCAD/FreeCAD/pull/8917